File: src/state.rs
use std::sync::atomic::{AtomicBool, AtomicU64};
use portable_atomic::AtomicF64;
use tokio::sync::Notify;

pub struct SharedState {
    pub mid_price: AtomicF64,
    pub price_timestamp_us: AtomicU64,
    pub feed_alive: AtomicBool,

    /// Feed signals the engine whenever a new price arrives.
    pub price_notify: Notify,

    pub cached_mid_ticks: AtomicU64,
    pub base_total_lots: AtomicU64,
    pub quote_total_lots: AtomicU64,
    pub onchain_sequence_number: AtomicU64,

    pub volatility_bps: AtomicF64,

    pub consecutive_failures: AtomicU64,

    pub cycles_total: AtomicU64,
    pub updates_sent: AtomicU64,
    pub mid_only_updates: AtomicU64,
    pub book_updates: AtomicU64,
    pub clear_book_sends: AtomicU64,
    pub heartbeat_sends: AtomicU64,

    /// MakerFillEvent stream metrics (populated by the fills subscriber).
    pub fills_count: AtomicU64,
    pub fill_base_lots: AtomicU64,
    pub fill_quote_lots: AtomicU64,
    /// Number of times inventory was refreshed from an on-chain book account update.
    pub book_resyncs: AtomicU64,

    pub engine_alive: AtomicBool,
}

impl SharedState {
    pub fn new() -> Self {
        Self {
            mid_price: AtomicF64::new(0.0),
            price_timestamp_us: AtomicU64::new(0),
            feed_alive: AtomicBool::new(false),
            price_notify: Notify::new(),
            cached_mid_ticks: AtomicU64::new(0),
            base_total_lots: AtomicU64::new(0),
            quote_total_lots: AtomicU64::new(0),
            onchain_sequence_number: AtomicU64::new(0),
            volatility_bps: AtomicF64::new(0.0),
            consecutive_failures: AtomicU64::new(0),
            cycles_total: AtomicU64::new(0),
            updates_sent: AtomicU64::new(0),
            mid_only_updates: AtomicU64::new(0),
            book_updates: AtomicU64::new(0),
            clear_book_sends: AtomicU64::new(0),
            heartbeat_sends: AtomicU64::new(0),
            fills_count: AtomicU64::new(0),
            fill_base_lots: AtomicU64::new(0),
            fill_quote_lots: AtomicU64::new(0),
            book_resyncs: AtomicU64::new(0),
            engine_alive: AtomicBool::new(false),
        }
    }
}

pub fn now_us() -> u64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or_default()
        .as_micros() as u64
}

File: ts/src/state.ts
// Port of src/state.rs. Rust used atomics only because tokio is multi-threaded;
// JS is single-threaded, so the shared state is a plain mutable object and the
// price Notify becomes a simple promise-based signal.

export function nowUs(): number {
  // Microseconds since the epoch. performance.timeOrigin + performance.now()
  // gives sub-millisecond resolution; Date.now() is the fallback.
  return Math.floor((performance.timeOrigin + performance.now()) * 1000);
}

// One-shot notifier the feed uses to wake the engine on each price tick.
// notifyOne() resolves any pending waiter; notified() resolves on the next
// notifyOne() after it is called.
export class Notify {
  private resolvers: (() => void)[] = [];

  notifyOne(): void {
    const r = this.resolvers.shift();
    if (r) r();
  }

  notified(): Promise<void> {
    return new Promise((resolve) => {
      this.resolvers.push(resolve);
    });
  }
}

export class SharedState {
  midPrice = 0;
  priceTimestampUs = 0;
  feedAlive = false;

  priceNotify = new Notify();

  cachedMidTicks = 0n;
  baseTotalLots = 0n;
  quoteTotalLots = 0n;
  onchainSequenceNumber = 0n;

  volatilityBps = 0;

  consecutiveFailures = 0;

  cyclesTotal = 0;
  updatesSent = 0;
  midOnlyUpdates = 0;
  bookUpdates = 0;
  clearBookSends = 0;
  heartbeatSends = 0;

  fillsCount = 0;
  fillBaseLots = 0n;
  fillQuoteLots = 0n;
  bookResyncs = 0;

  engineAlive = false;
}
